File: Swift/BankaiCore/Sources/BankaiCore/RepositoryFlow.swift
//
//  RepoSnapshotFlow.swift
//  
//
//  Created by Sergio Daniel on 1/09/24.
//

import Combine
import Foundation

public enum RepoSyncError: Error {
    case failedStoring(originalError: Error)
    case failedPushing(originalError: Error)
    
    case failedRetrieving(originalError: Error)
    case failedPulling(originalError: Error)
}

enum RepoSnapshot<T> {
    // Happy paths
    case local(T)
    case remote(T, String?)
}

public protocol SnapshotReceiver {
    associatedtype Data
    
    func send(local data: Data)
    func send(remote data: Data, from remoteName: String?)
    func giveUp()
    func fail(with error: RepoSyncError)
}

/// Represents a flow of snapshots of the given data type. Intended
/// to retrieve data from a local source expecting a potential remote
/// snapshot to contrast data and decide which one to keep.
/// - Parameters:
///     - Data: The type of data to manage
public class RepositoryFlow<Data>: SnapshotReceiver {
    // MARK: - Subtypes
    typealias Receiver = RepositoryFlow
    typealias Operation = (Receiver) async throws -> Void
    
    // MARK: - Instance Members
    
    let onlyLocalExpected: Bool
    
    
    // Whether the flow has completed already or not.
    var hasCompleted: Bool = false
    
    /// The operation block to be run when the flow is started.
    /// The body of the operation to be performed by the flow which should
    /// deliver values across time in order to flow to successfully complete.
    private var body: (Receiver) async throws -> Void = { _ in }
    
    private let scheduler: AnySchedulerOf<DispatchQueue>
    
    /// Subject holding flow of snapshots over time until it completes.
    private let subject = SubjectOf<RepoSnapshot<Data>, RepoSyncError>()
    
    private var _cancellable: AnyCancellable!
    
    /// Retrieves cancellable instance of the flow and registers handlers
    /// for each snapshot type.
    /// It also keeps an instance of cancellable in order to control flow
    /// from inside.
    func run() -> AnyCancellable {
        _cancellable = subject
            .receive(on: scheduler)
            .print("zheref")
            .subscribe(on: DispatchQueue.main)
            .handleEvents(receiveSubscription: { _ in
                Task { try await self.body(self) }
            }, receiveOutput: { snapshot in
                print(">>> Received snapshot")
                switch snapshot {
                case .local(let data):
                    print(">>> Flow detected incoming local data")
                    self.yieldLocal(data)
                case .remote(let data, let remoteName):
                    self.yieldRemote(data, remoteName)
                }
            }, receiveCompletion: { completion in
                print(">>> Received completion")
                self.hasCompleted = true
                switch completion {
                case .finished:
                    self.yieldCompletion()
                case .failure(let error):
                    self.yieldFailure(error)
                }
            })
            .sink { _ in } receiveValue: { _ in }
        return _cancellable
    }
    
    var yieldLocal: (Data) -> Void = { _ in }
    var yieldRemote: (Data, String?) -> Void = { _, _ in }
    var yieldFailure: (RepoSyncError) -> Void = { _ in }
    var yieldCompletion: () -> Void = { }
    
    // MARK: Constructors
    
    /// Creates a new instance of RepositoryFlow.
    /// Registers the operation to be run when the flow is started.
    /// This operation block will be responsible for emitting new snapshots
    /// or events to the flow.
    /// - Parameters:
    ///     - onlyLocalExpected: Whether the flow will only expect a 
    ///     local snapshot or also expect remote snapshots.
    ///     - scheduler: The scheduler where the flow should operate.
    ///     - operation: Async operation block taking a receiver handler
    ///     to send new snapshots/events with.
    init(onlyLocalExpected: Bool = false,
         scheduler: AnySchedulerOf<DispatchQueue> = .global(),
         _ operation: @escaping Operation) {
        self.onlyLocalExpected = onlyLocalExpected
        self.scheduler = scheduler
        self.body = operation
    }
    
    // MARK: - Sending values
    
    /// Sends a local snapshot to the flow so that subscribers are notified.
    /// - Parameters:
    ///     - local: The snapshot of data to send to the flow as local.
    public func send(local data: Data) {
        guard !hasCompleted else { return }
        print(">>> Will deliver local data")
        subject.send(.local(data))
        if onlyLocalExpected {
            subject.send(completion: .finished)
        }
    }
    
    /// Sends a remote snapshot to the flow so that subscribers are notified.
    /// - Parameters:
    ///     - remote: The value to be received by the flow.
    ///     - from: The name of the remote sending the new value.
    public func send(remote data: Data, from remoteName: String? = nil) {
        guard !hasCompleted else { return }
        subject.send(.remote(data, remoteName))
        subject.send(completion: .finished)
    }
    
    public func giveUp() {
        guard !hasCompleted else { return }
        subject.send(completion: .finished)
    }
    
    public func fail(with error: RepoSyncError) {
        guard !hasCompleted else { return }
        subject.send(completion: .failure(error))
    }
    
    // Subscribe to events
    
    @discardableResult
    func onLocal(_ yield: @escaping (Data) -> Void) -> Self {
        self.yieldLocal = yield
        return self
    }
    
    @discardableResult
    func onRemote(_ yield: @escaping (Data, String?) -> Void) -> Self {
        self.yieldRemote = yield
        return self
    }
    
    @discardableResult
    func onFailure(_ yield: @escaping (RepoSyncError) -> Void) -> Self {
        self.yieldFailure = yield
        return self
    }
    
    @discardableResult
    func onCompletion(_ yield: @escaping () -> Void) -> Self {
        self.yieldCompletion = yield
        return self
    }
}

File: C#/BankaiCore/BankaiCore/Repository/RepositoryFlow.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Reactive;
using System.Reactive.Concurrency;
using System.Reactive.Linq;
using System.Reactive.Subjects;
using Microsoft.VisualBasic.FileIO;

namespace BankaiCore.Repository;

#region Support Types

public class RepoSyncException(string message) : Exception(message) { }

public abstract record class RepoSnapshot<T>
{
    public sealed record Local(T data) : RepoSnapshot<T>;

    public sealed record Remote(T data, string? remoteName)
        : RepoSnapshot<T>;
}

/// <summary>
/// Represents an object you can send snapshots to.
/// </summary>
/// <typeparam name="T">The type of object you expect to send</typeparam>
public interface SnapshotReceiver<in T>
{
    /// <summary>
    /// Sends a local snapshot to the flow so that subscribers are notified.
    /// </summary>
    /// <param name="value">The value to be received by the flow</param>
    void sendLocal(T value);

    /// <summary>
    /// Sends a remote snapshot to the flow so that subscribers are notified.
    /// </summary>
    /// <param name="value">The value to be received by the flow</param>
    /// <param name="remoteName">The name of the remote sending the new value</param>
    /// <param name="shouldHold">Determines whether the flow should hold its completion or not</param>
    void sendRemote(T value, string? remoteName, bool shouldHold = false);
    void giveUp();
    void fail(RepoSyncException e);
}

#endregion

/// <summary>
/// Represents a flow of snapshots of the given data type. Intended
/// to retrieve data from a local source expecting a potential remote
/// snapshot to contrast data and decide which one to keep.
/// </summary>
/// <typeparam name="Data">The type of data to manage</typeparam>
public class RepositoryFlow<Data>: SnapshotReceiver<Data>
{
    #region Subtypes
    public delegate Task Operation(SnapshotReceiver<Data> receiver);
    #endregion

    private bool onlyLocalExpected { get; set; }

    /// <summary>
    /// Subject holding flow of snapshots over time until it completes.
    /// </summary>
    private readonly ReplaySubject<RepoSnapshot<Data>> subject = new();

    /// <summary>
    /// Whether the flow has completed already or not.
    /// </summary>
    public bool hasCompleted { get; set; }

    private event Action<Data> yieldLocal = data => { };
    private event Action<Data, string?> yieldRemote = (data, s) => { };
    private event Action<RepoSyncException> yieldFailure = e => { };
    private event Action yieldCompletion = () => { };

    /// <summary>
    /// The operation block to be run when the flow is started.
    /// The body of the operation to be performed by the flow which should
    /// deliver values across time in order to flow to successfully complete.
    /// </summary>
    private event Operation body = receiver => Task.Delay(0);

    private readonly IScheduler _scheduler;
    private IDisposable _disposable;

    /// <summary>
    /// Retrieves cancellable instance of the flow and registers handlers
    /// for each snapshot type.
    /// </summary>
    public IDisposable run()
    {
        var disposable = subject
            //.ObserveOn(_scheduler ?? Scheduler.Default)
            .SubscribeOn(_scheduler ?? Scheduler.Default)
            .Subscribe(handleSnapshot, exception =>
            {
                this.yieldFailure(
                    new RepoSyncException(exception.Message)
                );
            }, () =>
            {
                Console.WriteLine($">>> Completed on Thread: {Thread.CurrentThread.Name}");
                this.hasCompleted = true;
                this.yieldCompletion();
            });
        _ = _scheduler!.Schedule(() =>
        {
            this.body(this);
        });
        this._disposable = disposable;
        return this._disposable;
    }

    private void handleSnapshot(RepoSnapshot<Data> snapshot)
    {
        Console.WriteLine($"Received new snapshot");
        switch (snapshot)
        {
            case RepoSnapshot<Data>.Local castedSnapshot:
                this.yieldLocal(castedSnapshot.data);
                break;
            case RepoSnapshot<Data>.Remote castedSnapshot:
                this.yieldRemote(
                    castedSnapshot.data,
                    castedSnapshot.remoteName
                );
                break;
        }
    }

    #region Constructors

    /// <summary>
    /// Creates a new instance of RepositoryFlow.
    /// Registers the operation to be run when the flow is started.
    /// This operation block will be responsible for emitting new snapshots
    /// or events to the flow.
    /// </summary>
    /// <param name="operation">Async operation block taking a receiver handler to send new snapshots/events with</param>
    /// <param name="onlyLocalExpected">Whether the flow will only expect a local snapshot or also expect remote snapshots</param>
    /// <param name="scheduler">The scheduler where the flow should operate</param>
    public RepositoryFlow(Operation operation, bool onlyLocalExpected = false, IScheduler? scheduler = null)
    {
        this.onlyLocalExpected = onlyLocalExpected;
        _scheduler = scheduler ?? Scheduler.Default;
        this.body = operation;
    }

    #endregion

    #region Sending values

    /// <summary>
    /// Sends a local snapshot to the flow so that subscribers are notified.
    /// </summary>
    /// <param name="value">The value to be received by the flow</param>
    public void sendLocal(Data value)
    {
        Console.WriteLine($"Received ${value} as local snapshot");
        if (hasCompleted) return;
        subject.OnNext(new RepoSnapshot<Data>.Local(value));
        if (onlyLocalExpected)
            subject.OnCompleted();
    }

    /// <summary>
    /// Sends a remote snapshot to the flow so that subscribers are notified.
    /// </summary>
    /// <param name="value">The value to be received by the flow</param>
    /// <param name="remoteName">The name of the remote sending the new value</param>
    /// <param name="shouldHold">Determines whether the flow should hold its completion or not</param>
    public void sendRemote(Data value, string? remoteName, bool shouldHold = false)
    {
        if (hasCompleted) return;
        subject.OnNext(new RepoSnapshot<Data>.Remote(value, remoteName));
        if (!shouldHold)
            subject.OnCompleted();
    }

    public void giveUp()
    {
        if (hasCompleted) return;
        subject.OnCompleted();
    }

    public void fail(RepoSyncException e)
    {
        if (hasCompleted) return;
        subject.OnError(e);
    }

    #endregion

    #region Event Subscriptions

    /// <summary>
    /// Subscribes action to be performed when a local snapshot is received.
    /// </summary>
    /// <param name="action">The action to be performed</param>
    /// <returns>The flow with the action subscribed</returns>
    public RepositoryFlow<Data> onLocal(Action<Data> action)
    {
        this.yieldLocal += action;
        return this;
    }

    public RepositoryFlow<Data> onRemote(Action<Data, string?> action)
    {
        this.yieldRemote += action;
        return this;
    }

    public RepositoryFlow<Data> onFailure(Action<RepoSyncException> action)
    {
        this.yieldFailure += action;
        return this;
    }

    public RepositoryFlow<Data> onCompletion(Action action)
    {
        this.yieldCompletion += action;
        return this;
    }

    #endregion
}
